site/content/docs/how_the_optimizing_compiler_works/backend.md

```diff
@@ -0,0 +1,504 @@
+
+`wazevoapi.PrintSSAToBackendIRLowering` prints the basic blocks with the
+lowered arch-specific instructions.
+
+## Register Allocation
+
+The register allocation phase is responsible for mapping the potentially
+infinite number of virtual registers to the real registers of the target
+architecture. Because the number of real registers is limited, the register
+allocation phase might need to "spill" some of the virtual registers to memory;
+that is, it might store their content, and then load them back into a register
+when they are needed.
+
+For a given function `f` the register allocation procedure
+`regalloc.Allocator.DoAllocation(f)` is implemented in sub-phases:
+
+- `livenessAnalysis(f)` collects the "liveness" information for each virtual
+  register. The algorithm is described in [Chapter 9.2 of The SSA
+Book][ssa-book].
+
+- `alloc(f)` allocates registers for the given function. The algorithm is
+  derived from [the Go compiler's
+allocator][go-regalloc]
+
+At the end of the allocation procedure, we also record the set of registers
+that are **clobbered** by the body of the function. A register is clobbered
+if its value is overwritten by the function, and it is not saved by the
+callee. This information is used in the finalization phase to determine which
+registers need to be spilled in the prologue. This is not strictly related
+to register allocation in a textbook meaning, but it is a necessary step
+for the finalization phase.
+
+### Liveness Analysis
+
+Intuitively, a variable or name binding can be considered _live_ at a certain
+point in a program, if its value will be used in the future.
+
+For instance:
+
+```
+1| int f(int x) {
+2|   int y = 2 + x;
+3|   int z = x + y;
+4|   return z;
+5| }
+```
+
+Variable `x` and `y` are both live at line 4, because they are used in the
+expression `x + y` on line 3; variable `z` is live at line 4, because it is
+used in the return statement.  However, variables `x` and `y` can be considered
+_not_ live at line 4 because they are not used anywhere after line 3.
+
+Statically, _liveness_ can be approximated by following paths backwards on the
+control-flow graph, connecting the uses of a given variable to its definitions
+(or its *unique* definition, assuming SSA form).
+
+In practice, while liveness is a property of each name binding at any point in
+the program, it is enough to keep track of liveness at the boundaries of basic
+blocks:
+
+- the _live-in_ set for a given basic block is the set of all bindings that are
+  live at the entry of that block.
+- the _live-out_ set for a given basic block is the set of all bindings that
+  are live at the exit of that block. A binding is live at the exit of a block
+if it is live at the entry of a successor.
+
+Because the CFG is a connected graph, it is enough to keep track of either
+live-in or live-out sets, and then propagate the liveness information backwards
+or forwards, respectively. In our case, we keep track of live-ins.
```

Comment:
would this work?

```suggestion
Because the CFG is a connected graph, it is enough to keep track of either
live-in or live-out sets, and then propagate the liveness information backward
or forward, respectively. In our case, we keep track of  live-in sets per block;
live-outs are derived from live-ins of the successor blocks when a block is allocated.
```

Reply:
works for me